returnn/datasets/map.py

```diff
@@ -0,0 +1,210 @@
+"""
+Provides :class:`MapDatasetBase`
+"""
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.datasets.cached2 import CachedDataset2
+from returnn.util.basic import OptionalNotImplementedError
+
+
+class MapDatasetBase(object):
+  """
+  This dataset can be used as template to implement user-side Datasets, where the data can be access in arbitrary order.
+  For global sorting, the length information needs to be known beforehand.
+  """
+
+  def __init__(self, num_outputs=None):
```

Comment:
What is `num_outputs`? Is this only about outputs? What is the dict content?

Reply:
This is called num_outputs in other places in the code, but yes, it would be good to change this misleading name here. I've now called it data_types lacking a better idea. Might be confused with data types as in float, int, etc. But alternatives would be also not very clear: data_specs, data_defs, data_config... :thinking:

Reply:
One of the main reasons for this new alternative interface `MapDatasetBase` is that this is supposed to be more straight-forward, easier, and can get rid of other confusing parts. It doesn't really makes sense to replicate the confusing parts.

Also, `num_outputs` is already deprecated. In the main config, we have `extern_data` instead, and in the `Dataset`, we should also introduce sth more generic. But anyway, that's all irrelevant, as with `MapDatasetBase`, you start completely fresh, and can design it in any way you want. Maybe similar to the PyTorch API, as some people might already know that.